src/porepy/viz/diagnostics_mixin.py

```diff
@@ -0,0 +1,383 @@
+            is_plotting_condition_number (optional): Whether to plot
+                the condition number. Causion - this might take time if the matrix
+                is big. Defaults to False.
+            is_plotting_max (optional): Whether to plot the absolute maximum.
+                This option is cheap even for big matrices.
+                Defaults to True.
+        """
+
+        def get_eig(mat, which) -> float:
+            return svds(mat, k=1, return_singular_vectors=False, which=which).item()
+
+        def get_condition_number(mat) -> float:
+            largest_eig = get_eig(mat, which="LM")
+            # Smallest eigenvalue might be zero.
+            smallest_eig = get_eig(mat, which="SM")
+            if mat.data.size == 0:
+                return 0.0
+            else:
+                return float(largest_eig / (smallest_eig + 1e-15))
+
+        full_matrix: csr_matrix = self.linear_system[0]
+        if not _IS_SEABORN_AVAILABLE:
+            logger.warning(
+                "Plotting the diagnostics image requires seaborn package."
+                ' Run "pip install seaborn". Falling back to text mode.'
+            )
+
+        if is_plotting_condition_number and full_matrix.shape[0] > 1e5:
+            logger.warning(
+                "Computing condition number might take significant time for "
+                "big matrices. It is recommended to reduce the problem size or to use "
+                'the option "is_plotting_max" only.'
+            )
+
+        # Determining the block indices and collecting the submatrices.
+        equation_indices = self._equations_indices(grouping=grouping)
+        variable_indices = self._variable_indices(grouping=grouping)
+
+        if grouping is None and max(len(equation_indices), len(variable_indices)) > 10:
+            logger.warning(
+                "Treating all the subdomains separately might not be "
+                "informative if there are many of them. It is recommended to set"
+                "\"grouping='subdomains'\""
+            )
+
+        self._collect_submatrices(
+            mat=full_matrix,
+            equation_indices=equation_indices,
+            variable_indices=variable_indices,
+        )
+
+        # Computing the required features of each block.
+        block_matrix_shape = len(equation_indices), len(variable_indices)
+        block_data: dict[tuple[int, int], dict] = defaultdict(dict)
+        for i, j in product(range(block_matrix_shape[0]), range(block_matrix_shape[1])):
+            submat = self.submatrices[i, j]
+            is_empty_block = submat.data[submat.data != 0].size == 0
+            block_data[i, j]["is_empty_block"] = is_empty_block
+            if not is_empty_block:
+                if is_plotting_condition_number:
+                    block_data[i, j]["condition_number"] = get_condition_number(submat)
+                if is_plotting_max:
+                    block_data[i, j]["max"] = abs(submat.data).max()
```

Comment:
I think you can expand the utility of this function by adding kwargs on the form dict[str, Callable[[Matrix], float]] and then at this point do something like

for key, func in kwargs.items()
    block_data[i, j][key] = func(submat)

If you also return block_data, the user can have increased flexibility in doing tailored diagnostics.